### PR TITLE
EBG-15851 update repair truncation handle when MP transactions are executed.

### PR DESCRIPTION
**Pull request by zeemanhe**
Repair truncation handles for MP transactions are set when they are pushed to task queue.  The handles could be way behind when these tasks are executed, which could lead to no-op in the truncation process and consume heap. 
- Update the tasks with latest truncation handle when they are executed.  
- advance the truncation handle for any completed transactions not just for committed transactions. 
- Capture the repair truncation handle upon MIP failover.

### CODE REVIEW

File: src/frontend/org/voltdb/iv2/MpInitiator.java [hunk 1]
```diff
@@ -133,15 +133,17 @@ public void acceptPromotion()
...
                     RepairResult res = repair.start().get();
                     txnid = res.m_txnId;
+                    repairTruncationHandle = res.m_repairTruncationHandle;
                     success = true;
                 } catch (CancellationException e) {
                     success = false;
                 }
                 if (success) {
-                    m_initiatorMailbox.setLeaderState(txnid);
+                    ((MpInitiatorMailbox)m_initiatorMailbox).setLeaderState(txnid, repairTruncationHandle);
```
Comment (rdykiel) [inline src/frontend/org/voltdb/iv2/MpInitiator.java]: Why can't this method be defined in the base class of the m_initiatorMailbox instance? Is it MPI-specific?

File: src/frontend/org/voltdb/iv2/MpScheduler.java [hunk 2]
```diff
@@ -172,8 +172,9 @@ public void shutdown()
                 VoltMessage resp = counter.getLastResponse();
                 if (resp != null && resp instanceof InitiateResponseMessage) {
                     InitiateResponseMessage msg = (InitiateResponseMessage)resp;
-                    if (msg.shouldCommit() && msg.haveSentMpFragment()) {
+                    if (msg.haveSentMpFragment()) {
                         m_repairLogTruncationHandle = m_repairLogAwaitingCommit;
+                        m_pendingTasks.setRepairLogTruncationHandle(m_repairLogTruncationHandle);
                         m_repairLogAwaitingCommit = msg.getTxnId();
```
Comment (rdykiel) [inline src/frontend/org/voltdb/iv2/MpScheduler.java]: Is this assignment still OK when the messages are not to be committed? What does it mean now "awaiting commit" in this context?

File: src/frontend/org/voltdb/iv2/MpScheduler.java [hunk 3]
```diff
@@ -536,9 +538,10 @@ else if (result == DuplicateCounter.MISMATCH) {
             // doing duplicate suppresion: all done.
         }
         else {
-            // Only advance the truncation point on committed transactions that sent fragments to SPIs.
-            if (message.shouldCommit() && message.haveSentMpFragment()) {
+            // Only advance the truncation point on completed transactions that sent fragments to SPIs.
+            if (message.haveSentMpFragment()) {
                 m_repairLogTruncationHandle = m_repairLogAwaitingCommit;
+                m_pendingTasks.setRepairLogTruncationHandle(m_repairLogTruncationHandle);
                 m_repairLogAwaitingCommit = message.getTxnId();
```
Comment (wweiss-voltdb) [inline src/frontend/org/voltdb/iv2/MpScheduler.java]: m_repairLogAwaitingCommit should be renamed.

File: src/frontend/org/voltdb/iv2/RepairLog.java [hunk 4]
```diff
@@ -58,6 +58,9 @@
     long m_lastSpHandle = Long.MAX_VALUE;
     long m_lastMpHandle = Long.MAX_VALUE;
 
+    // The MP repair truncation handle
+    long m_repairTruncationHandle = Long.MIN_VALUE;
```
Comment (luyangco) [inline src/frontend/org/voltdb/iv2/RepairLog.java]: It's better to rename the variable with `mp` prefix.

File: src/frontend/org/voltdb/iv2/MpProcedureTask.java [hunk 5]
```diff
@@ -182,6 +182,11 @@ public void run(SiteProcedureConnection siteConnection)
         if (hostLog.isDebugEnabled()) {
             hostLog.debug("[MpProcedureTask] STARTING: " + this);
         }
+
+        final long truncationHandle = Math.max(((MpTransactionTaskQueue)m_queue).getRepairLogTruncationHandle(),
```
Comment (luyangco) [inline src/frontend/org/voltdb/iv2/MpProcedureTask.java]: I think the reason to update truncationHandle MpTransactionTaskQueue is for restarted MP txn to have a new truncation handle, if you agree with me,  could you add a comment to show your intent?

**Comment by luyangco** [comment]
If a misrouted MP read-write TXN needs to restart, does its truncationHandle gets updated with new value?

File: src/frontend/org/voltdb/iv2/MpScheduler.java [hunk 6]
```diff
@@ -559,6 +547,16 @@ else if (result == DuplicateCounter.MISMATCH) {
         }
     }
 
+    private void advanceRepairTruncationHandle(InitiateResponseMessage msg) {
+        // Only advance the truncation point on completed transactions that sent fragments to SPIs.
+        // See ENG-4211 & ENG-14563
+        if(msg.shouldCommit() && msg.haveSentMpFragment()) {
```
Comment (luyangco) [inline src/frontend/org/voltdb/iv2/MpScheduler.java]: TruncationHandle is still set by committed transactions, not completed transactions (e.g. aborted transaction), it doesn't reflect the Pull Request description, is this your intention?

Comment (zeemanhe) [reply src/frontend/org/voltdb/iv2/MpScheduler.java]: decided to address this issue in a separate ticket.

**Pull request merged**